src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+        }
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_147: [ execute_callbacks shall perform cleanup of the operation. ]*/
+        THANDLE_ASSIGN(RC_PTR)(&channel_op->data, NULL);
+        THANDLE_ASSIGN(CHANNEL_INTERNAL)(&channel_op->channel_internal, NULL);
+
+        // copy to local reference to avoid cutting the branch you are sitting on
+        THANDLE(ASYNC_OP) temp = NULL;
+        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(&temp, &channel_op->async_op);
+        THANDLE_ASSIGN(ASYNC_OP)(&temp, NULL);
+    }
+}
+
+static void cancel_op(void* context)
+{
+    CHANNEL_OP* channel_op = context;
+    CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_op->channel_internal);
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_134: [ cancel_op shall call srw_lock_acquire_exclusive. ]*/
+    srw_lock_acquire_exclusive(channel_internal_ptr->lock);
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_135: [ If the operation is in the list of pending operations: ]*/
+        if (channel_op->anchor.Flink != &channel_op->anchor && &channel_op->anchor == channel_op->anchor.Flink->Blink)
+        {
+            /*Codes_SRS_CHANNEL_INTERNAL_43_137: [ cancel_op shall call DList_RemoveEntryList to remove the operation from the list of pending operations. ]*/
+            (void)DList_RemoveEntryList(&channel_op->anchor);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_136: [ cancel_op shall set the result of the operation to CHANNEL_CALLBACK_RESULT_CANCELLED. ]*/
+            channel_op->result = CHANNEL_CALLBACK_RESULT_CANCELLED;
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_138: [ cancel_op shall call threadpool_schedule_work with execute_callbacks as work_function and the operation as work_function_context. ]*/
+            if (threadpool_schedule_work(channel_internal_ptr->threadpool, execute_callbacks, channel_op) != 0)
+            {
+                LogError("Failure in threadpool_schedule_work(execute_callbacks, channel_op)");
+            }
+        }
+        /*Codes_SRS_CHANNEL_INTERNAL_43_146: [ If the operation is not in the list of pending operations and the result of the operation is CHANNEL_CALLBACK_RESULT_OK, cancel_op shall set the result of the operation to CHANNEL_CALLBACK_RESULT_CANCELLED. ]*/
+        else if (channel_op->result == CHANNEL_CALLBACK_RESULT_OK)
+        {
+            channel_op->result = CHANNEL_CALLBACK_RESULT_CANCELLED;
+        }
+    }
+    /*Codes_SRS_CHANNEL_INTERNAL_43_139: [ cancel_op shall call srw_lock_release_exclusive. ]*/
+    srw_lock_release_exclusive(channel_internal_ptr->lock);
+}
+
+static void dispose_channel_op(void* context)
+{
+    /* don't need to do anything here because actual cleanup happens in execute_callbacks */
+    (void)context;
+}
+
+static int enqueue_operation(CHANNEL_INTERNAL* channel_internal, THANDLE(ASYNC_OP)* out_op, PULL_CALLBACK pull_callback, void* pull_context, PUSH_CALLBACK push_callback, void* push_context, THANDLE(RC_PTR) data)
```

Comment:
>enqueue_operation [](http://example.com/codeflow?start=11&length=17)

I don't like this function. It is misleading.

I believe it is misleading because it populates the fields and enqueues. However, it only ever builds HALF of the operations (either the pull half or the push half).

At any time when this function is called, either PUSH or PULL callback are going to be NULL.

Reply:
Each pull/push is only half an operation. `enqueue_operation` does half the work. What you have described is precisely what I intend it to do. 

The alternative would be to write the same code for pull  and push, with the only difference being which fields are set.